src/qx/QuantumState.cpp

```diff
@@ -0,0 +1,81 @@
+#include <fmt/format.h>
+
+#include "qx/QuantumState.hpp"
+
+
+namespace qx::core {
+
+void QuantumState::checkQuantumState() {
+    if (numberOfQubits == 0) {
+        throw QuantumStateError{ "number of qubits needs to be at least 1" };
+    }
+    if (numberOfQubits > config::MAX_QUBIT_NUMBER) {
+        throw QuantumStateError{ fmt::format("number of qubits exceeds maximum allowed: {} > {}", numberOfQubits,
+                                             config::MAX_QUBIT_NUMBER) };
+    }
+    if (not isNormalized()) {
+        throw QuantumStateError{ "quantum state is not normalized" };
+    }
+}
+
+QuantumState::QuantumState(std::size_t qubit_register_size)
+    : numberOfQubits{ qubit_register_size }
+    , data{ static_cast<size_t>(1) << numberOfQubits } {
+
+    data[BasisVector{}] = SparseComplex{ 1. };  // start initialized in state 00...000
+    checkQuantumState();
+}
+
+QuantumState::QuantumState(std::size_t qubit_register_size,
+                           std::initializer_list<std::pair<std::string, std::complex<double>>> values)
+    : numberOfQubits{ qubit_register_size }
+    , data{ static_cast<size_t>(1) << numberOfQubits, values } {
+
+    checkQuantumState();
+}
+
+[[nodiscard]] std::size_t QuantumState::getNumberOfQubits() const {
+    return numberOfQubits;
+}
+
+[[nodiscard]] bool QuantumState::isNormalized() {
+    return isNull(data.norm() - 1.);
+}
+
+void QuantumState::reset() {
```

Comment:
We will need to investigate this a bit further because the reset instruction of a single qubit will effect qubits around it if it  is entangled.